src/main/java/jewellery/inventory/service/ProductService.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+    public List<ProductResponseDto> getAllProducts() {
+        List<Product> products = productRepository.findAll();
+        return products.stream().map(productMapper::mapToProductResponseDto).toList();
+    }
+
+    public ProductResponseDto getProduct(UUID id) {
+        Product product = productRepository.findById(id)
+                .orElseThrow(() -> new ProductNotFoundException(id));
+
+        return productMapper.mapToProductResponseDto(product);
+    }
+
+    public void deleteProduct(UUID id) {
+
+        Product product = productRepository.findById(id).orElseThrow(
+                () -> new ProductNotFoundException(id));
+
+        throwExceptionIfProductIsSold(id, product);
+        throwExceptionIfProductIsPartOfAnotherProduct(id, product);
+
+        moveResourceInProductToResourceInUser(product);
+        disassembleProductContent(product);
+
+        productRepository.deleteById(id);
+    }
+
+    private void throwExceptionIfProductIsPartOfAnotherProduct(UUID id, Product product) {
+        if (product.getContentOf() != null) {
+            throw new ProductIsContentException(id);
+        }
+    }
+
+    private void throwExceptionIfProductIsSold(UUID id, Product product) {
+        if (product.isSold()) {
+            throw new ProductIsSoldException(id);
+        }
+    }
+
+    private void disassembleProductContent(Product product) {
+        if (product.getProductsContent() != null) {
+            product.getProductsContent().forEach(content -> {
+                content.setContentOf(null);
+                productRepository.save(content);
+            });
+
+            product.setProductsContent(new ArrayList<>());
+            productRepository.save(product);
+        }
+    }
+
+    private void moveResourceInProductToResourceInUser(Product product) {
+        List<ResourceInProduct> resourcesInProduct = product.getResourcesContent();
+        User owner = product.getOwner();
+
+        List<ResourceInUser> resultResourcesInUser = new ArrayList<>();
+
+        resourcesInProduct.forEach(resourceInProduct -> {
+            Resource resource = resourceInProduct.getResource();
+            ResourceInUser resourceInUser = moveResourcesToUserWhenDisassembleProduct(owner, resourceInProduct, resource);
+
+            resultResourcesInUser.add(resourceInUser);
+
+            if (resourceInProduct.getProduct() == null) {
+                resourceInProductRepository.deleteById(resourceInProduct.getId());
+            }
+        });
+
+        resourceInUserRepository.saveAll(resultResourcesInUser);
+    }
+
+    private User getUserFromRequest(ProductRequestDto productRequestDto) {
+        return userRepository.findById(productRequestDto.getOwnerId())
+                .orElseThrow(() -> new UserNotFoundException(productRequestDto.getOwnerId()));
+    }
+
+    private List<ResourceInProduct> getResourcesFromRequest(User user, List<ResourceQuantityRequestDto> resourcesInProductRequestDto) {
```

Comment:
This method is doing a lot more than just getting the resources from a request. It is also removing quantities from user and adding them to products. This functionality shouldn't be part of this method.
Let's discuss how we can improve this together.